File: project_euler/Problem053/Problem053.py
class Problem053:
    def solution(self) -> None:
        target, result = 1000000, 0
        for i in range(23, 101):
            result += i + 1 - (self.__get_index(i, target) << 1)
        print(result)

    def __get_index(self, value: int, target: int) -> int:
        start, end = 0, (value + 1) >> 1
        while start <= end:
            current = start + ((end - start) >> 1)
            if self.__check_combination(value, current, target):
                start = current + 1
            else:
                end = current - 1
        return start

    def __check_combination(self, value: int, count: int, target: int) -> bool:
        result = 1
        for i in range(count):
            result = result * (value - i) // (i + 1)
            if result > target:
                return False
        return True

File: project_euler/Problem053/Problem053.java
package project_euler;

public class Problem053 {
    public void solution() {
        int target = 1000000;
        int result = 0;
        for (int i = 23; i <= 100; ++i) {
            result += i + 1 - (getIndex(i, target) << 1);
        }
        System.out.println(result);
    }

    private int getIndex(int value, int target) {
        int start = 0;
        int end = (value + 1) >> 1;
        while (start <= end) {
            int current = start + ((end - start) >> 1);
            if (checkCombination(value, current, target)) {
                start = current + 1;
            } else {
                end = current - 1;
            }
        }
        return start;
    }

    private boolean checkCombination(int value, int count, int target) {
        int result = 1;
        for (int i = 0; i < count; ++i) {
            result = result * (value - i) / (i + 1);
            if (result > target) {
                return false;
            }
        }
        return true;
    }
}
